### ISSUE TITLE
NestedSet - deleteNode : You should not use beforeUpdate?

### ISSUE THREAD

**Issue by fenixphp** [phalcon/incubator#603]
``` php
<?php namespace fastPanel\Navigate\Models\Menu;

/**
 * Link.php
 *
 * @author    Fenixphp <fenixphp@gmail.com>
 * @copyright 2014 - 2016 Desionlab
 * @license   MIT
 */

use Phalcon\Mvc\Model;
use Phalcon\Mvc\Model\Behavior\NestedSet;
use Phalcon\Mvc\Model\Behavior\Timestampable;

/**
 * Class Link
 *
 * Set description and links of navigational menu.
 *
 * @package fastPanel\Navigate\Models\Menu
 * @version 1.0.0
 */
class Link extends Model {

  /**
   * Element identifier.
   *
   * @var int
   */
  public $id;

  /**
   * Kind accessory element.
   *
   * @var string
   */
  public $kind;

  /**
   * The key to the menu record.
   *
   * @var int
   */
  public $menu_id;

  /**
   * Nested set "root" attribute.
   *
   * @var int
   */
  public $root;

  /**
   * Nested set "left" attribute.
   *
   * @var int
   */
  public $lft;

  /**
   * Nested set "right" attribute.
   *
   * @var int
   */
  public $rgt;

  /**
   * Nested set "level" attribute.
   *
   * @var int
   */
  public $level;

  /**
   * Title (name) of the element.
   *
   * @var string
   */
  public $title;

  /**
   * Short description.
   *
   * @var string
   */
  public $description;

  /**
   * URL string to follow a link.
   *
   * @var string
   */
  public $href;

  /**
   * A set of classes to display graphical icons.
   *
   * @var string
   */
  public $icon;

  /**
   * A set of classes for css.
   *
   * @var string
   */
  public $css;

  /**
   * Direction to follow a link.
   *
   * @var string
   */
  public $target;

  /**
   * Date creation of the element.
   *
   * @var \DateTime
   */
  public $created_at;

  /**
   * Date change of the element.
   *
   * @var \DateTime
   */
  public $updated_at;

  /**
   * Status of the enabled element.
   *
   * @var int
   */
  public $enabled;

  /**
   * Init.
   */
  public function initialize() {
    /* Set table. */
    $this->setSource('navigate_menu_links');

    /* Logger date. */
    $this->addBehavior(new Timestampable([
      'beforeCreate' => [
        'field'  => ['created_at', 'updated_at'],
        'format' => 'Y-m-d H:i:s'
      ],
      'beforeUpdate' => [
        'field'  => 'updated_at',
        'format' => 'Y-m-d H:i:s'
      ]
    ]));

    /* Nested set logic. */
    $this->addBehavior(new NestedSet([
      'hasManyRoots' => true
    ]));

    /* Menu. */
    $this->belongsTo(
      'menu_id',
      'fastPanel\Navigate\Models\Menu',
      'id',
      [
        'alias'  => 'menu'
      ]
    );
  }

  /**
   * Callback.
   */
  public function beforeCreate() {

  }

  /**
   * Callback.
   */
  public function beforeDelete() {
    /*  */
    if ($this->kind == 'system') {
      return false;
    }

    return true;
  }

  /**
   * Callback.
   */
  public function beforeSave() {

  }

  /**
   * Callback.
   */
  public function afterFetch() {

  }

}

/* End of file Link.php */
```

Everything works except deleteNode methods. What could I do wrong?

**Comment by fenixphp** [comment]
```php

    /* Find record. */
    $record = \fastPanel\Navigate\Models\Menu\Link::findFirst($id);

    /* Check record. */
    if (!$record) {
      /* Send not found. */
      return $this->response
        ->setStatusCode(404, 'Not found.')
          ->setJsonContent(JsonSend::failed(404, 'Not found.', []));
    }

    /* Remove from database. */
    if ($record->deleteNode()) {
      /* Send data. */
      return $this->response
        ->setStatusCode(200, 'Ok.')
          ->setJsonContent(JsonSend::success(200, 'Ok.', []));
    } else {
      /* Send errors messages. */
      return $this->response
        ->setStatusCode(422, 'Unprocessed entity.')
          ->setJsonContent(JsonSend::failedValidate(422, 'Unprocessed entity.', $record->getMessages()));
    }
```

Error: You should not use fastPanel\Navigate\Models\Menu\Link:beforeUpdate when Phalcon\Mvc\Model\Behavior\NestedSet attached. Use the methods of behavior.

What is wrong here? Thanks in advance.

**Comment by sergeyklay** [comment]
What version of Incubator?

**Comment by fenixphp** [comment]
2.1.x

**Comment by sergeyklay** [comment]
I'll try to sort out

**Comment by fenixphp** [comment]
Ok .. Thank.

**Comment by gorellnet** [comment]
This is still happening in phalcon 3.0.0 with the latest incubator from master. Any news/schedule when it will be fixed?

**Comment by sergeyklay** [comment]
It's not Phalcon issue

**Comment by alekciy** [comment]
OC: Debian 7, PHP: 5.4.45, Phalcon: 2.0.10, phalcon/incubator: ^2.0
deleteNode() not work.

```
You should not use app\models\Segment:beforeUpdate when Phalcon\Mvc\Model\Behavior\NestedSet attached. Use the methods of behavior.

#0 [internal function]: Phalcon\Mvc\Model\Behavior\NestedSet->notify('beforeUpdate', Object(app\models\Segment))
#1 [internal function]: Phalcon\Mvc\Model\Manager->notifyEvent('beforeUpdate', Object(app\models\Segment))
#2 [internal function]: Phalcon\Mvc\Model->fireEventCancel('beforeUpdate')
#3 [internal function]: Phalcon\Mvc\Model->_preSave(Object(Phalcon\Mvc\Model\MetaData\Memory), true, 'id')
#4 [internal function]: Phalcon\Mvc\Model->save(Array, NULL)
#5 /var/www/http/vendor/phalcon/incubator/Library/Phalcon/Mvc/Model/Behavior/NestedSet.php(779): Phalcon\Mvc\Model->update(Array)
#6 /var/www/http/vendor/phalcon/incubator/Library/Phalcon/Mvc/Model/Behavior/NestedSet.php(586): Phalcon\Mvc\Model\Behavior\NestedSet->shiftLeftRight(9, -6)
#7 [internal function]: Phalcon\Mvc\Model\Behavior\NestedSet->deleteNode()
#8 /var/www/http/vendor/phalcon/incubator/Library/Phalcon/Mvc/Model/Behavior/NestedSet.php(108): call_user_func_array(Array, Array)
#9 [internal function]: Phalcon\Mvc\Model\Behavior\NestedSet->missingMethod(Object(app\models\Segment), 'deleteNode', Array)
#10 [internal function]: Phalcon\Mvc\Model\Manager->missingMethod(Object(app\models\Segment), 'deleteNode', Array)
#11 /var/www/http/app/controllers/SegmentController.php(85): Phalcon\Mvc\Model->__call('deleteNode', Array)
#12 /var/www/http/app/controllers/SegmentController.php(85): app\models\Segment->deleteNode()
#13 [internal function]: app\controllers\SegmentController->batchAction()
#14 [internal function]: Phalcon\Dispatcher->dispatch()
#15 /var/www/http/public/index.php(16): Phalcon\Mvc\Application->handle()
#16 {main}
```

NestedSet use [update](https://github.com/phalcon/incubator/blob/master/Library/Phalcon/Mvc/Model/Behavior/NestedSet.php#L779) method. How fix?

**Comment by rabrowne85** [comment]
Looking at some of the other functions the shiftLeftRight function usually has an ignoreEvent call before/after it. Therefore, wrapping ```$this->shiftLeftRight($key, $delta);``` on Line 586 with the ignoreEvent call as follows seems to solve the issue:
```
$this->ignoreEvent = true;
$this->shiftLeftRight($key, $delta);
$this->ignoreEvent = false;
```

N.B. I have no idea if this is the full/valid solution, but it is definitely a "workaround".

**Comment by 3amprogrammer** [comment]
@rabrowne85 every other `shiftLeftRight` method call across NestedSet class is wrapped with ignoreEvent assignments so this seems like a valid solution.

**Comment by sergeyklay** [comment]
Fixed in the `3.2.x branch`. Feel free to open new issue if the problem appears again. Thank you for contributing.
